pkg/core/tekton/builder/pipeline.go

```diff
@@ -0,0 +1,131 @@
+package builder
+
+import (
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PipelineBuilder holds the tekton pipeline definition.
+type PipelineBuilder struct {
+	Pipeline v1beta1.Pipeline
+}
+
+// NewPipelineBuilder creates a Pipeline with default values.
+func NewPipelineBuilder(name, namespace string) *PipelineBuilder {
+	b := &PipelineBuilder{}
+	b.Pipeline = v1beta1.Pipeline{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+	return b
+}
+
+// PipelineMeta sets the Meta structs of the Pipeline.
+// Any number of MetaOp modifiers can be passed.
+func (b *PipelineBuilder) PipelineMeta(ops ...MetaOp) {
```

Comment:
Really awesome use of functional optional here ! Reviewing your code is a real learning experience for me, thank you !

Have you thought about using the same pattern with the `v1beta1.Pipeline` object ? For example, this very function could just be another functional optional, this time applied to the pipeline object, like so:

```
type PipelineOption func(*v1beta1.Pipeline)

func WithMeta(ops ...MetaOp) PipelineOption {
	return func(pipeline *v1beta1.Pipeline) {
	    for _, op := range ops {
		    switch o := op.(type) {
		    case ObjectMetaOp:
			    o(&pipeline.ObjectMeta)
		    case TypeMetaOp:
			    o(&pipeline.TypeMeta)
		    }
	    }
      }
}
```

Of course, in this case we're talking about more than just a pipeline constructor, but I don't see why you wouldn't be able to use functional options with an pipeline update function as well.